src/ring_theory/dedekind_domain/factorization.lean

```diff
@@ -0,0 +1,196 @@
+`v^-(val_v(I))` is not the unit ideal. -/
+lemma finite_mul_support_inv {I : ideal R} (hI : I ≠ 0) :
+  (mul_support (λ (v : height_one_spectrum R),
+    (v.as_ideal : fractional_ideal (non_zero_divisors R) K)^
+      -((associates.mk v.as_ideal).count (associates.mk I).factors : ℤ))).finite :=
+begin
+  rw mul_support,
+  simp_rw [zpow_neg, ne.def, inv_eq_one],
+  exact ideal.finite_mul_support_coe hI,
+end
+
+/-- For every nonzero ideal `I` of `v`, `v^(val_v(I) + 1)` does not divide `∏_v v^(val_v(I))`. -/
+lemma finprod_not_dvd (I : ideal R) (hI : I ≠ 0) :
+  ¬ (v.as_ideal)^((associates.mk v.as_ideal).count (associates.mk I).factors + 1) ∣
+      (∏ᶠ (v : height_one_spectrum R), (v.as_ideal)^
+        (associates.mk v.as_ideal).count (associates.mk I).factors) :=
+begin
+  have hf := ideal.finite_mul_support hI,
+  have h_ne_zero : v.as_ideal ^
+    (associates.mk v.as_ideal).count (associates.mk I).factors ≠ 0 := pow_ne_zero _ v.ne_bot,
+  rw [← mul_finprod_cond_ne v hf, pow_add, pow_one, finprod_cond_ne _ _ hf],
+  intro h_contr,
+  have hv_prime : prime v.as_ideal := ideal.prime_of_is_prime v.ne_bot v.is_prime,
+  obtain ⟨w, hw, hvw'⟩ :=
+  prime.exists_mem_finset_dvd hv_prime ((mul_dvd_mul_iff_left h_ne_zero).mp h_contr),
+  have hw_prime : prime w.as_ideal := ideal.prime_of_is_prime w.ne_bot w.is_prime,
+  have hvw := prime.dvd_of_dvd_pow hv_prime hvw',
+  rw [prime.dvd_prime_iff_associated hv_prime hw_prime, associated_iff_eq] at hvw,
+  exact (finset.mem_erase.mp hw).1 (height_one_spectrum.ext w v (eq.symm hvw)),
+end
+
+end ideal
+
+lemma associates.finprod_ne_zero (I : ideal R) :
+  associates.mk (∏ᶠ (v : height_one_spectrum R), v.as_ideal ^
+    (associates.mk v.as_ideal).count (associates.mk I).factors) ≠ 0 :=
+begin
+  rw [associates.mk_ne_zero, finprod_def],
+  split_ifs,
+  { rw finset.prod_ne_zero_iff,
+    intros v hv,
+    apply pow_ne_zero _ v.ne_bot, },
+  { exact one_ne_zero, }
+end
+
+namespace ideal
+
+/-- The multiplicity of `v` in `∏_v v^(val_v(I))` equals `val_v(I)`. -/
+lemma finprod_count (I : ideal R) (hI : I ≠ 0) :
+  (associates.mk v.as_ideal).count (associates.mk (∏ᶠ (v : height_one_spectrum R), (v.as_ideal)^
+    (associates.mk v.as_ideal).count (associates.mk I).factors)).factors =
+    (associates.mk v.as_ideal).count (associates.mk I).factors :=
+begin
+  have h_ne_zero := associates.finprod_ne_zero I,
+  have hv : irreducible (associates.mk v.as_ideal) := v.associates_irreducible,
+  have h_dvd := finprod_mem_dvd _ (ideal.finite_mul_support hI),
+  have h_not_dvd := ideal.finprod_not_dvd v I hI,
+  rw [← associates.mk_dvd_mk, associates.dvd_eq_le, associates.mk_pow,
+    associates.prime_pow_dvd_iff_le h_ne_zero hv] at h_dvd h_not_dvd,
+  rw not_le at h_not_dvd,
+  apply nat.eq_of_le_of_lt_succ h_dvd h_not_dvd,
+end
+
+/-- The ideal `I` equals the finprod `∏_v v^(val_v(I))`. -/
+lemma factorization (I : ideal R) (hI : I ≠ 0) : ∏ᶠ (v : height_one_spectrum R), (v.as_ideal)^
```

Comment:
I don't like this name. Since there are many ways to factor an ideal we should be more specific.
```suggestion
/-- The ideal `I` equals the finprod `∏_v v^(val_v(I))`. -/
lemma finprod_max_pow_dividing (I : ideal R) (hI : I ≠ 0) : ∏ᶠ (v : height_one_spectrum R), (v.as_ideal)^
```

Reply:
What about `irreducible_factorization` (or maybe `prime_factorization`)? Since we are working over a Dedekind domain, this is unique up to reordering.

Reply:
You're right of course that factorizations are unique! Let me try to word it better. I think there are sufficiently many different ways of expressing the factorization of an ideal in Lean that it warrants a more specific name for this result: we could write the product using list product, multiset products, finset products or `finprod`, the irreducibles could be `height_one_spectrum` or `irreducible` or `prime` or `ideal.prime`. By choosing a specific name we can avoid confusion between all the different, still equivalent, statements.

By the way, since the mathlib docs also search in docstrings, you could add something like `/-- **Factorization of ideals**: The ideal `I` equals the finprod `∏_v v^(val_v(I))`. -/` in the docstring, then someone who searches the docs for `ideal factorization` will find this result.

Reply:
How about `finprod_factorization` then? I think that having factorization in the name makes the Lean code more readable for mathematicians, but this would still specify in which way the factorization is implemented.

Reply:
Okay, I can live with factorization in the name. Ideally I would also like something to indicate it concerns the `height_one_spectrum` rather than some other way of denoting the prime ideals. So maybe `finprod_height_one_spectrum_factorization`? But I think this has been sitting long enough that it's fine with either name.

Reply:
OK, I can rename it to `finprod_height_one_spectrum_factorization`.